### PR TITLE
[feat] - Introduce `channelmetrics` Package for Channel Metrics Collection

### PR DESCRIPTION
**Pull request by ahrav**
### Description:
This PR introduces a new `channelmetrics` package that provides functionality to collect and record various metrics related to channel operations. This enhancement will help in monitoring the performance and utilization of Go channels, allowing for better optimization and troubleshooting.

### Additions

1. **MetricsCollector:**
   - Interface for recording channel metrics:
     - `produceDuration`: Histogram for production time.
     - `consumeDuration`: Histogram for consumption time.
     - `channelLen`: Gauge for current buffer size.
     - `channelCap`: Gauge for buffer capacity.

2. **NewMetricsCollector:**
   - Creates a `MetricsCollector` with metrics organized by `namespace`, `subsystem`, and `chanName`.
   - Ensures a `MetricsCollector` is initialized only once per unique combination of `chanName`, `namespace`, and `subsystem` using a map to track instances and `sync.Once` for thread-safe initialization.

3. **ObservableChan:**
   - Wraps a Go channel to collect metrics:
     - `Close()`: Records buffer size on close.
     - `Send(ctx, item)`: Records send duration and buffer size.
     - `Recv(ctx)`: Records receive duration and buffer size.

### Example Metrics
Screenshot 2024-05-27 at 12 16 28 PM

### Checklist:
* [ ] Tests passing (`make test-community`)?
* [ ] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/usage/install/#local-installation))?

### CODE REVIEW

File: pkg/channelmetrics/observablechan.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+// Close closes the channel and records the current size of the channel buffer.
+func (oc *ObservableChan[T]) Close() {
+	close(oc.ch)
+	oc.RecordChannelLen()
+}
+
+// Send sends an item into the channel and records the duration taken to do so.
+// It also updates the current size of the channel buffer.
+func (oc *ObservableChan[T]) Send(_ context.Context, item T) {
```
Comment (ahrav) [inline pkg/channelmetrics/observablechan.go:53]: ctx is unused for now, same for `Recv`

**Comment by dustin-decker** [comment]
This is cool!

File: pkg/channelmetrics/observablechan.go (hunk 1 above)
Comment (dustin-decker) [reply pkg/channelmetrics/observablechan.go:53]: Combine with CancelableWrite? https://github.com/trufflesecurity/trufflehog/blob/024b219dfbb18442d71da5c1bdfcdc93920e41d0/pkg/common/context.go#L17

**Review by dustin-decker — APPROVED** [review]
Good stuff, I think we should include cancelable write and receive though.

**Comment by ahrav** [comment]
> Good stuff, I think we should include cancelable write and receive though.

Yea, I like that idea. Thanks.

File: pkg/channelmetrics/observablechan.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	RecordChannelCap(capacity int)
+}
+
+// ObservableChan wraps a Go channel and collects metrics about its usage.
+// It supports any type of channel and records metrics using a provided
+// MetricsCollector implementation.
+type ObservableChan[T any] struct {
+	ch        chan T
+	metrics   MetricsCollector
+	bufferCap int
```
Comment (mcastorina) [inline pkg/channelmetrics/observablechan.go:28]: Why is this its own struct member? Should it always be equal to `cap(ch)`?

File: pkg/channelmetrics/observablechan.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+// using the provided MetricsCollector and channel name. The chanName is used to
+// distinguish between metrics for different channels by incorporating it into
+// the metric names.
+func NewObservableChan[T any](ch chan T, metrics MetricsCollector) *ObservableChan[T] {
+	oChan := &ObservableChan[T]{
+		ch:        ch,
+		metrics:   metrics,
+		bufferCap: cap(ch),
+	}
+	oChan.RecordChannelCapacity() // Record capacity immediately
```
Comment (mcastorina) [inline pkg/channelmetrics/observablechan.go:42]: Should we also record the length immediately? In most cases it'll be 0, but theoretically any channel can be converted into an `ObservableChan`.

File: pkg/channelmetrics/observablechan.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+// Recv receives an item from the channel and records the duration taken to do so.
+// It also updates the current size of the channel buffer.
+func (oc *ObservableChan[T]) Recv(_ context.Context) T {
+	startTime := time.Now()
+	defer func() {
+		oc.metrics.RecordConsumeDuration(time.Since(startTime))
+		oc.RecordChannelLen()
+	}()
```
Comment (mcastorina) [inline pkg/channelmetrics/observablechan.go:72]: **Nit:**

```go
defer func(start time.Time) {
    // ...
}(time.Now())
```

File: pkg/channelmetrics/observablechan.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	return <-oc.ch
```
Comment (mcastorina) [inline pkg/channelmetrics/observablechan.go:73]: Cancellable recv?

File: pkg/channelmetrics/metrics_collector/prometheus/collector.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+// you can identify bottlenecks, optimize performance, and ensure that the ObservableChan is operating efficiently.
+func NewMetricsCollector(chanName, namespace, subsystem string) *MetricsCollector {
+	key := fmt.Sprintf("%s_%s_%s", namespace, subsystem, chanName)
+
+	collectorsMu.Lock()
+	defer collectorsMu.Unlock()
+
+	if collector, exists := collectors[key]; exists {
+		return collector
+	}
```
Comment (mcastorina) [inline pkg/channelmetrics/metrics_collector/prometheus/collector.go:59]: Is there ever a case where we want `NewMetricsCollector` to return an existing collector? Or would this indicate we have a name collision?

Comment (ahrav) [reply pkg/channelmetrics/metrics_collector/prometheus/collector.go:59]: That would result in a collision, correct.

File: pkg/channelmetrics/observablechan.go (hunk 3 above)
Comment (ahrav) [reply pkg/channelmetrics/observablechan.go:42]: I didn't account for that use case, thanks.

File: pkg/channelmetrics/observablechan.go (hunk 5 above)
Comment (ahrav) [reply pkg/channelmetrics/observablechan.go:73]: Do we have an existing CancellableRecv, or are you suggesting I implement it here?

Comment (mcastorina) [reply pkg/channelmetrics/observablechan.go:73]: No existing one that I know of.

File: pkg/channelmetrics/metrics_collector/prometheus/collector.go (hunk 6 above)
Comment (mcastorina) [reply pkg/channelmetrics/metrics_collector/prometheus/collector.go:59]: Is a collision something that should never happen?

**Pull request merged**